plugin/v2/plugin.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Implementation of the KMS Plugin API v2.
+package v2
+
+import (
+	"regexp"
+
+	"google.golang.org/api/cloudkms/v1"
+
+	"context"
+	"encoding/base64"
+	"fmt"
+	"time"
+
+	"github.com/GoogleCloudPlatform/k8s-cloudkms-plugin/plugin"
+	"github.com/golang/glog"
+)
+
+const (
+	apiVersion      = "v2beta1"
+	ok              = "ok"
+	ping            = "cGluZw=="
+	keyNotReachable = "Cloud KMS key is not reachable"
+	keyDisabled     = "Cloud KMS key is not enabled or no cloudkms.cryptoKeys.get permission"
+)
+
+// Store the last known primary key version resource name
+// to return as KeyId in case when the Cloud KMS service is not reachable
+// because KeyId in StatusResponse cannot be empty and shouldn't trigger key migration
+// in case of transient remote service unavailability
+var lastKeyId string
+
+// Regex to extract Cloud KMS key resource name from the key version resource name
+var keyResourceRegEx = regexp.MustCompile(`projects\/[^/]+\/locations\/[^/]+\/keyRings\/[^/]+\/cryptoKeys\/[^/:]+`)
+
+type Plugin struct {
+	*plugin.AbstractPlugin
+}
+
+// New constructs Plugin.
+func NewPlugin(keyService *cloudkms.ProjectsLocationsKeyRingsCryptoKeysService, keyURI, keySuffix string, pathToUnixSocketFile string) *Plugin {
+	lastKeyId = getKeyId(keyURI, keySuffix)
+	p := &Plugin{
+		AbstractPlugin: &plugin.AbstractPlugin{
+			KeyService:       keyService,
+			KeyURI:           keyURI,
+			KeySuffix:        keySuffix,
+			PathToUnixSocket: pathToUnixSocketFile,
+		},
+	}
+	p.AbstractPlugin.Plugin = p
+	return p
+}
+
+// Status returns the version of KMS API version that plugin supports.
+// Response also contains the status of the plugin, which is calculated as availability of the
+// encryption key that the plugin is confinged with, and the current primary key version.
+// kube-apiserver will provide this key version in Encrypt and Decrypt calls and will be able
+// to know whether the remote CLoud KMS key has been rotated or not.
+func (g *Plugin) Status(ctx context.Context, request *StatusRequest) (*StatusResponse, error) {
+	defer plugin.RecordCloudKMSOperation("encrypt", time.Now().UTC())
+
+	var response StatusResponse
+	req := &cloudkms.EncryptRequest{Plaintext: ping}
```

Comment:
You probably don't want to do this on every request. 

Perhaps maintain a lastStatus + timestamp which is updated every $duration. 
You only need to trigger an explicit 'ping' if the newest response is too stale.

Reply:
The period for calling the Status method determines Kubernetes.
The API server polls the Status procedure call approximately every minute when everything is healthy, and every 10 seconds when the plugin is not healthy.
As I learned from Di Jin, that is what the GKE implementation doing. Using paid encrypt/decrypt call pair instead of a free get call has advantages: 1) no additional IAM permissions required 2) the status check is very accurate.
Calling Cloud KMS API twie in a minute should not be a big overhead.